crates/stark-backend/src/prover/quotient/evaluator.rs

```diff
@@ -0,0 +1,167 @@
+        match self {
+            PackedExpr::Val(x) => PackedExpr::Val(-x),
+            PackedExpr::Challenge(x) => PackedExpr::Challenge(-x),
+        }
+    }
+}
+
+impl<SC> SymbolicEvaluator<Val<SC>, PackedExpr<SC>> for ProverConstraintEvaluator<'_, SC>
+where
+    SC: StarkGenericConfig,
+{
+    fn eval_const(&self, c: Val<SC>) -> PackedExpr<SC> {
+        PackedExpr::Val(c.into())
+    }
+    fn eval_is_first_row(&self) -> PackedExpr<SC> {
+        PackedExpr::Val(self.is_first_row)
+    }
+    fn eval_is_last_row(&self) -> PackedExpr<SC> {
+        PackedExpr::Val(self.is_last_row)
+    }
+    fn eval_is_transition(&self) -> PackedExpr<SC> {
+        PackedExpr::Val(self.is_transition)
+    }
+
+    fn eval_var(&self, symbolic_var: SymbolicVariable<Val<SC>>) -> PackedExpr<SC> {
+        let index = symbolic_var.index;
+        match symbolic_var.entry {
+            Entry::Preprocessed { offset } => PackedExpr::Val(self.preprocessed.get(offset, index)),
+            Entry::Main { part_index, offset } => {
+                PackedExpr::Val(self.partitioned_main[part_index].get(offset, index))
+            }
+            Entry::Public => PackedExpr::Val(self.public_values[index].into()),
+            Entry::Permutation { offset } => {
+                let perm = self
+                    .after_challenge
+                    .first()
+                    .expect("Challenge phase not supported");
+                PackedExpr::Challenge(perm.get(offset, index))
+            }
+            Entry::Challenge => {
+                let permutation_randomness = self
+                    .challenges
+                    .first()
+                    .map(|c| c.as_slice())
+                    .expect("Challenge phase not supported");
+                PackedExpr::Challenge(permutation_randomness[index])
+            }
+            Entry::Exposed => {
+                let permutation_exposed_values = self
+                    .exposed_values_after_challenge
+                    .first()
+                    .expect("Challenge phase not supported");
+                PackedExpr::Challenge(permutation_exposed_values[index])
+            }
+        }
+    }
+}
+
+impl<SC: StarkGenericConfig> ProverConstraintEvaluator<'_, SC> {
+    /// `alpha_powers` are in **reversed** order, with highest power coming first.
+    // Note: this could be split into multiple functions if additional constraints need to be folded in
+    pub fn accumulate(
+        &self,
+        constraints: &SymbolicExpressionDag<Val<SC>>,
+        alpha_powers: &[PackedChallenge<SC>],
+    ) -> PackedChallenge<SC> {
+        let evaluated_nodes = self.eval_nodes(&constraints.nodes);
+        let mut accumulator = PackedChallenge::<SC>::ZERO;
+        for (&alpha_pow, &node_idx) in alpha_powers.iter().zip(&constraints.constraint_idx) {
+            match evaluated_nodes[node_idx] {
+                PackedExpr::Val(x) => {
+                    accumulator += alpha_pow * x;
+                }
```

Comment:
nit that I thought `cargo fmt` would catch
```suggestion
                PackedExpr::Val(x) => accumulator += alpha_pow * x,
```

Reply:
interesting, indeed it doesn't catch but also does accept